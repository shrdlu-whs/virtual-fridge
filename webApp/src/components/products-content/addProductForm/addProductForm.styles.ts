import { createStyles, lighten, makeStyles, Theme } from "@material-ui/core/styles";

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    ingredientButton: {
      marginTop: "16px",
      marginBottom: "8px",
    },

    addProductContentContainer: {
      padding: theme.spacing(5, 5),
    },

    addProductTitle: {
      marginTop: 0,
      textAlign: "center",
      marginBottom: theme.spacing(3),
      textTransform: "uppercase",
      paddingBottom: theme.spacing(2),
      borderBottom: `1px solid ${theme.palette.primary.main}`,
    },

    categorylist:{
      width: '100%',
      maxWidth: 360,
      backgroundColor: theme.palette.background.paper,
    },

    icon_expand: {
      float: 'right',
      transform: 'rotate(0deg)',
      marginLeft: 'auto',
      transition: theme.transitions.create('transform', {
        duration: theme.transitions.duration.shortest,
      }),
    },

    icon_expandOpen: {
      float: 'right',
      transform: 'rotate(180deg)',
    },

    formControl: {
      margin: theme.spacing(1),
      minWidth: 120,
    },
  })
);
