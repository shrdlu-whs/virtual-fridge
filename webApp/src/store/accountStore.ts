/* eslint-disable no-debugger */
import { makeObservable, action, computed, observable } from "mobx";
import { Auth } from "aws-amplify";
import { RootStore } from "./rootStore";
import {
  IRegisterForm,
  ISignInForm,
  IForgotPasswordForm,
  IForgotPasswordSubmitForm,
  IUserInfo,
} from "../interfaces/account.interface";
import axios, { Canceler, CancelToken, CancelTokenSource, CancelTokenStatic } from "axios";


// Amplify: https://docs.amplify.aws/lib/auth/getting-started/q/platform/js

export class AccountStore {
  @observable userInfo: IUserInfo | null = null;
  @observable forgotUsername: string | undefined = undefined;

  // cancelToken: CancelTokenSource = axios.CancelToken.source()
  cancelRequest: Canceler | undefined = undefined;

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  generateCancelToken() {
    return new axios.CancelToken((c) => {
      this.cancelRequest = c;
    })
  }

  @action
  async register(register: IRegisterForm) {
    try {
      await Auth.signUp({
        username: register.username,
        password: register.password,
        attributes: {
          email: register.email,
        },
      });
      this.rootStore.routeStore.redirect("/signIn");
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "register",
        "error",
        true,
        error
      );
    }
  }

  @action
  async signIn(signIn: ISignInForm) {
    try {
      this.rootStore.uiStateStore.setSubmitting(true);
      await Auth.signIn(signIn.username, signIn.password);
      await this.fetchUserInfo();
      this.rootStore.routeStore.redirect("/products");
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "signIn",
        "error",
        true,
        error
      );
    } finally {
      this.rootStore.uiStateStore.setSubmitting(false);
    }
  }

  @action
  async signOut() {
    try {
      await Auth.signOut();
      // this.userInfo = null; // If the site refresh, fetchUserInfo() will be executed and the userInfo will set anyway
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "signOut",
        "error",
        true,
        error
      );
    } finally {
      this.rootStore.routeStore.redirect("/signIn");
    }
  }

  @action
  async forgotPassword({ username }: IForgotPasswordForm) {
    try {
      await Auth.forgotPassword(username);
      this.setForgotUsername(username);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "forgot_password",
        "error",
        true,
        error
      );
    }
  }

  @action
  async forgotPasswordSubmit({
    username,
    code,
    password,
  }: IForgotPasswordSubmitForm) {
    try {
      await Auth.forgotPasswordSubmit(username, code, password);
      this.rootStore.routeStore.redirect("/signIn");
      this.setForgotUsername(undefined);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "forgot_password_submit",
        "error",
        true,
        error
      );
    }
  }

  @action
  async changePassword(oldPassword: string, newPassword: string) {
    try {
      const user = await Auth.currentAuthenticatedUser();
      await Auth.changePassword(user, oldPassword, newPassword);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "change_password",
        "error",
        true,
        error
      );
    }
  }

  @action
  async fetchUserInfo() {
    if (!this.rootStore.uiStateStore.fetchingUser) {
      try {
        this.rootStore.uiStateStore.setFetchingUser(true);
        const userInfo = (await Auth.currentUserInfo()) as IUserInfo | null;
        // const userInfo: IUserInfo = {username: "Test", attributes: {email: "test@mail.com", email_verified: "true", sub: "" }};
        this.setUserInfo(userInfo);
      } catch (error) {
        this.setUserInfo(null);
        this.rootStore.notificationStore.createNotification(
          "fetch_user",
          "error",
          true,
          error
        );
      } finally {
        this.rootStore.uiStateStore.setFetchingUser(false);
        // Its initiating, also when the an error occured => the user returns to login page
        if(!this.rootStore.uiStateStore.isInitiating) {
          this.rootStore.uiStateStore.setInitiated(true);
        }
      }
    }
  }

  @action
  setUserInfo(userInfo: IUserInfo | null) {
    this.userInfo = userInfo;
  }

  @action
  setForgotUsername(username: string | undefined) {
    this.forgotUsername = username;
  }
}
