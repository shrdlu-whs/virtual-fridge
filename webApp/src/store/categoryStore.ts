/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { IApiCategory } from "../interfaces/api/api-category.interface";
import { getCategories } from "../services/category2.service";

export class CategoryStore {
  @observable categories: IApiCategory[] = [];

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getCategories() {
    try {
      const categories = (await getCategories()).category;
      this.setCategories(categories);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_categories", "error", true, error);
      }
    }
  }

  @action
  setCategories(categories: IApiCategory[]) {
    this.categories = categories;
  }
}
