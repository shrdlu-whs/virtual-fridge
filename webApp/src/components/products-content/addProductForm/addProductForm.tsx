import React, { useEffect } from "react";
import * as Yup from "yup";
import clsx from "clsx";
import { runInAction } from "mobx";
import { Formik } from "formik";
import { IApiProduct } from "../../../interfaces/api/api-product.interface";
import { IApiProductItem } from "../../../interfaces/api/api-productItem.interface";
import { IApiCategory } from "../../../interfaces/api/api-category.interface";
import { ProductStore } from "../../../store/productStore";
import { CategoryStore } from "../../../store/categoryStore";
import { UiStateStore } from "../../../store/uiStateStore";
import { useTranslation } from "react-i18next";
import { useStyles } from "./addProductForm.styles";
import RequestButton from "../../request-button/requestButton";
import ExitToApp from "@material-ui/icons/ExitToApp";
import {
  FormControl,
  Toolbar,
  Button,
  Chip,
  Grid,
  Paper,
  TextField,
  Tooltip,
  Typography,
  List,
  ListItem,
  ListItemIcon,
  ListItemSecondaryAction,
  ListItemText,
  Checkbox,
  IconButton,
  Collapse,
  InputLabel,
  Select,
} from "@material-ui/core";
import { ExpandMore } from "@material-ui/icons";
import { inject, observer } from "mobx-react";
import {
  ICategory,
  INutritionValue,
  IProductItem,
  IProduct,
} from "../../../interfaces/froms/addProductForm.interface";
import { Category } from "@material-ui/icons";
import AddProductItemForm from "./addProductItemForm/addProductItemForm";
import ProductItemsTable from "./productItemsTable/productItemsTable";

interface IAddProductFormProps {
  productStore?: ProductStore;
  uiStateStore?: UiStateStore;
  categoryStore?: CategoryStore;
  isSubmitting: boolean;
  isMobile: boolean;
  handleAddProduct: (product: IApiProduct) => void;
  handleAddProductCancel: () => void;
}

export const AddProductForm = ({
  productStore,
  uiStateStore,
  categoryStore,
  isSubmitting,
  isMobile,
  handleAddProduct,
  handleAddProductCancel,
}: IAddProductFormProps) => {
  const [categories, setCategories] = React.useState<IApiCategory[]>([]);
  const { t } = useTranslation();
  const classes = useStyles();

  React.useEffect(() => {
    // categoryStore!.getCategories();
  }, []);

  const validateProduct = (formData: IProduct) => {
    const product: IApiProduct = {
      name: formData.name,
      iconPath: "",
      category: formData.category.map(
        (category): IApiCategory => {
          return { id: category.id, name: category.name };
        }
      ),
      unitDefaultQuantity: 100,
      totalQuantity: formData.totalQuantity,
      minQuantity: -1,
      unit: formData.unit,
      items: formData.items.map(
        (item): IApiProductItem => {
          return {
            barcode: item.barcode,
            expirationDate: new Date(item.expirationDate),
          };
        }
      ),
      nutritionalValue: formData.nutritionalValue,
    };
    product!.nutritionalValue!.nutritionScore = nutriState;
    handleAddProduct(product);
    productStore!.resetAddProductFormData();
  };

  const handleAddProductCancelClick = () => {
    productStore!.resetAddProductFormData();
    handleAddProductCancel();
  };

  const handleAddProductItem = (productItem: IProductItem) => {
    runInAction(() => productStore!.addProductFormData.items.push(productItem));
  };

  const handleDeleteProductItems = (uuids: string[]) => {
    for (const uuid of uuids) {
      const index = productStore!.addProductFormData.items.findIndex(
        (item) => item.uuid === uuid
      );
      if (index !== -1) {
        runInAction(() =>
          productStore!.addProductFormData.items.splice(index, 1)
        );
      }
    }
  };

  const handleUpdateProductItems = (
    item: IProductItem,
    barcode: string,
    expirationDate: Date
  ) => {
    runInAction(
      () => (
        (item.barcode = barcode),
        (item.expirationDate = expirationDate.toString())
      )
    );
  };

  const [catexpanded, setCatExpanded] = React.useState(false);
  const [itemexpanded, setItemExpanded] = React.useState(false);

  const handleCatExpandClick = () => {
    setCatExpanded(!catexpanded);
  };

  const handleItemExpandClick = () => {
    setItemExpanded(!itemexpanded);
  };

  const handleToggle = (value: number) => () => {
    let category = categoryStore!.categories.find(
      (element) => element.id === value
    );
    let index = productStore!.addProductFormData.category.findIndex(
      (x) => x.id === category!.id
    );
    if (index !== -1)
      productStore!.addProductFormData.category.splice(index, 1);
    else productStore!.addProductFormData.category.push(category!);
  };

  const [nutriState, setNutriState] = React.useState("");

  const handleNutriChange = (
    event: React.ChangeEvent<{
      name?: string | undefined;
      value: unknown;
    }>
  ) => {
    const val = event.target.value as string;
    setNutriState(val);
  };

  return (
    <Grid container direction="row" justify="center" alignItems="center">
      <Grid item sm={9} md={5} lg={4} xl={3}>
        <Paper className={classes.addProductContentContainer}>
          <Typography
            className={classes.addProductTitle}
            component="h1"
            variant="h5"
          >
            {t("title.product_add_product_header")}
          </Typography>
          <Formik
            initialValues={productStore!.addProductFormData}
            onSubmit={(values) => {
              validateProduct(values);
            }}
            enableReinitialize
          >
            {({
              values,
              touched,
              errors,
              dirty,
              handleChange,
              handleBlur,
              handleSubmit,
              isValid,
            }) => {
              return (
                <form onSubmit={handleSubmit}>
                  <div>
                    <TextField
                      label={t("label.product_name")}
                      name="name"
                      type="text"
                      variant="standard"
                      value={values.name}
                      onChange={handleChange}
                      onBlur={handleBlur}
                      helperText={touched.name && errors.name}
                      error={touched.name && errors.name !== undefined}
                      margin="normal"
                      fullWidth
                      required
                    />
                    <Grid container spacing={2}>
                      <Grid item xs={6}>
                        <TextField
                          label={t("label.product_totalQuantity")}
                          name="totalQuantity"
                          type="number"
                          variant="standard"
                          value={values.totalQuantity}
                          onChange={handleChange}
                          onBlur={handleBlur}
                          helperText={
                            touched.totalQuantity && errors.totalQuantity
                          }
                          error={
                            touched.totalQuantity &&
                            errors.totalQuantity !== undefined
                          }
                          margin="normal"
                          fullWidth
                        />
                      </Grid>
                      <Grid item xs={6}>
                        <TextField
                          label={t("label.product_unit")}
                          name="unit"
                          type="text"
                          variant="standard"
                          value={values.unit}
                          onChange={handleChange}
                          onBlur={handleBlur}
                          helperText={touched.unit && errors.unit}
                          error={touched.unit && errors.unit !== undefined}
                          margin="normal"
                          multiline
                          fullWidth
                        />
                      </Grid>
                    </Grid>
                  </div>
                  <Grid container spacing={2}>
                    <Grid container item xs={6} alignItems="center">
                      {t("label.product_categories_title")}
                    </Grid>
                    <Grid item xs={6}>
                      <IconButton
                        className={clsx(classes.icon_expand, {
                          [classes.icon_expandOpen]: catexpanded,
                        })}
                        onClick={handleCatExpandClick}
                        aria-expanded={catexpanded}
                        aria-label="show more"
                      >
                        <ExpandMore />
                      </IconButton>
                    </Grid>
                  </Grid>
                  <Collapse in={catexpanded} timeout="auto" unmountOnExit>
                    <List className={classes.categorylist}>
                      {categoryStore!.categories.map((category) => {
                        const labelId = `checkbox-list-label-${category.id}`;

                        return (
                          <ListItem
                            key={category.id}
                            role={undefined}
                            dense
                            button
                            onClick={handleToggle(category.id)}
                          >
                            <ListItemIcon>
                              <Checkbox
                                edge="start"
                                // checked={checked.indexOf(category.id) !== -1}
                                tabIndex={-1}
                                disableRipple
                                inputProps={{ "aria-labelledby": labelId }}
                              />
                            </ListItemIcon>
                            <ListItemText
                              id={labelId}
                              primary={category.name}
                            />
                          </ListItem>
                        );
                      })}
                    </List>
                  </Collapse>
                  <hr />
                  <Grid container spacing={2}>
                    <Grid container item xs={6} alignItems="center">
                      {t("title.product_additems_title")}
                    </Grid>
                    <Grid item xs={6}>
                      <IconButton
                        className={clsx(classes.icon_expand, {
                          [classes.icon_expandOpen]: itemexpanded,
                        })}
                        onClick={handleItemExpandClick}
                        aria-expanded={itemexpanded}
                        aria-label="show more"
                      >
                        <ExpandMore />
                      </IconButton>
                    </Grid>
                  </Grid>
                  <Collapse in={itemexpanded} timeout="auto" unmountOnExit>
                    <ProductItemsTable
                      handleDeleteProductItems={handleDeleteProductItems}
                      handleUpdateProductItems={handleUpdateProductItems}
                      isMobile={isMobile}
                      rows={values.items}
                    />
                    <AddProductItemForm
                      products={productStore!.products}
                      handleAddProductItem={handleAddProductItem}
                    />
                  </Collapse>
                  <hr />
                  <Grid container spacing={2}>
                    <Grid container item xs={12} alignItems="center">
                      {t("label.product_values_in_hundrets")}
                    </Grid>
                    <Grid item xs={12}>
                      <FormControl className={classes.formControl}>
                        <InputLabel htmlFor="nutriScore">
                          {t("label.product_nutriScore")}
                        </InputLabel>
                        <Select
                          native
                          value={nutriState}
                          onChange={handleNutriChange}
                          inputProps={{
                            name: "nutritionalValue.nutritionScore",
                            id: "nutritionalValue.nutritionScore",
                          }}
                        >
                          <option aria-label="None" value="" />
                          <option value={"A"}>A</option>
                          <option value={"B"}>B</option>
                          <option value={"C"}>C</option>
                          <option value={"D"}>D</option>
                          <option value={"E"}>E</option>
                        </Select>
                      </FormControl>
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_calories")}
                        name="nutritionalValue.calories"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_protein")}
                        name="nutritionalValue.protein"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"                        
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_carbohydrates")}
                        name="nutritionalValue.carbohydrates"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"                        
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_sugar")}
                        name="nutritionalValue.sugar"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_fat")}
                        name="nutritionalValue.fat"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_saturatedFat")}
                        name="nutritionalValue.saturatedFat"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_fiber")}
                        name="nutritionalValue.fiber"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={6}>
                      <TextField
                        label={t("label.product_salt")}
                        name="nutritionalValue.salt"
                        type="number"
                        variant="standard"
                        onChange={handleChange}
                        onBlur={handleBlur}
                        helperText={touched.unit && errors.unit}
                        error={touched.unit && errors.unit !== undefined}
                        margin="normal"
                        fullWidth
                      />
                    </Grid>
                  </Grid>
                  <Grid container spacing={2}>
                    <Grid item xs={6}>
                      <Button
                        onClick={handleAddProductCancelClick}
                        color="primary"
                      >
                        {t("action.cancel")}
                      </Button>
                    </Grid>
                    <Grid item xs={6}>
                      <RequestButton
                        variant="contained"
                        color="primary"
                        fullWidth
                        type="submit"
                        loading={isSubmitting}
                        disabled={
                          !touched || isSubmitting || (dirty && !isValid)
                        }
                        endIcon={<ExitToApp />}
                      >
                        {t("action.add")}
                      </RequestButton>
                    </Grid>
                  </Grid>
                </form>
              );
            }}
          </Formik>
        </Paper>
      </Grid>
    </Grid>
  );
};

export default inject(
  "productStore",
  "uiStateStore",
  "categoryStore"
)(observer(AddProductForm));
