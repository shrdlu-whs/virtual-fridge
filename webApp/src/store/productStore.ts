/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { IApiProduct } from "../interfaces/api/api-product.interface";
import { IIronStockProduct } from "../interfaces/ironStockProduct.interface";
import { getProducts, getProduct, addProduct, updateProduct, deleteProduct } from "../services/product2.service";
import { IApiCategory } from "../interfaces/api/api-category.interface";
import { INutritionValue, IProductItem, IProduct } from "../interfaces/froms/addProductForm.interface";
import { EMPTY_OBJECT } from "mobx/dist/internal";

export class ProductStore {

  nutritionValueEmpty: INutritionValue = {
    calories: 0,
    carbohydrates: 0,
    fat: 0,
    fiber: 0,
    nutritionScore: "A",
    protein: 0,
    salt: 0,
    saturatedFat: 0,
    sugar: 0
  };

  @observable activeProduct: IApiProduct | undefined = undefined;
  @observable products: IApiProduct[] = [];
  @observable ironStockProducts: IIronStockProduct[] = [];
  @observable missingIronStockProducts: IIronStockProduct[] = [];
  @observable productItems: IProductItem[] = [];

  @observable addProductFormData: IProduct = {
    name: "",
    iconPath: "",
    category: [],
    totalQuantity: 1,
    minQuantity: -1,
    unit: "",
    items: this.productItems,
    nutritionalValue: this.nutritionValueEmpty
  };

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getProducts() {
    try {
      // const products = (await getProducts(this.rootStore.accountStore.generateCancelToken())).data;
      this.rootStore.uiStateStore.setLoading(true);
      const products = (await getProducts()).products;
      this.setProducts(products);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_products", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async getProduct(id: number) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      // const product = (await getProduct(id, this.rootStore.accountStore.generateCancelToken())).data;
      const product = (await getProduct(id)).product;
      this.setActiveProduct(product);
      this.updateCachedProducts(product);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_product", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async addProduct(product: IApiProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      product = (await addProduct(product)).createProduct;
      this.setProducts(this.products.concat([product]));
    } catch (error) {
      console.log(error)
      this.rootStore.notificationStore.createNotification("add_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async updateProduct(product: IApiProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      await updateProduct(product);
      if (product.id) this.updateCachedProducts(product);
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async deleteProduct(id: number) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      await deleteProduct(id);
      const index = this.products.findIndex((product) => product.id === id);
      if (index !== -1) {
        runInAction(() => {
          this.products.splice(index, 1);
        });
      }
    } catch (error) {
      this.rootStore.notificationStore.createNotification("delete_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  @action
  updateCachedProducts(product: IApiProduct) {
    // Update cached products
    const index = this.products.findIndex((sProduct) => sProduct.id === product.id);
    if (index !== -1) {
      this.products.splice(index, 1, product);
    } else {
      this.products.push(product);
    }
  }

  async getIronStockProducts() {
    try {
      const products = (await getProducts()).products;
      this.rootStore.uiStateStore.setLoading(true);
      this.setProducts(products);
      const ironStockProducts = products.map(
        (product): IIronStockProduct => {
          return { id: product.id!, name: product.name, stock: product.totalQuantity, ironStock: product.minQuantity };
        }
      );
      this.setIronStockProducts(ironStockProducts.filter((product) => product.ironStock >= 0));
      this.setMissingIronStockProducts(ironStockProducts.filter((product) => product.ironStock < 0));
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_products", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async addIronStockProduct(ironStockProduct: IIronStockProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      const cachedProduct = this.products.find((sProduct) => sProduct.id === ironStockProduct.id)!;
      cachedProduct.minQuantity = ironStockProduct.ironStock;
      await updateProduct(cachedProduct);
      this.setIronStockProducts(this.ironStockProducts.concat([ironStockProduct]));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async updateIronStockProducts(ids: number[]) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      for (const id of ids) {
        const product = this.ironStockProducts.find((product) => product.id === id);
        if (product) {
          const cachedProduct = this.products.find((sProduct) => sProduct.id === product.id)!;
          cachedProduct.minQuantity = product.ironStock;
          await updateProduct(cachedProduct);
        }
      }
    } catch (error) {
      // Rollback ???
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  @action
  updateCachedIronStockProduct(id: number, ironStock: number) {
    const ironStockProduct = this.ironStockProducts.find((product) => product.id === id);
    if (ironStockProduct) {
      ironStockProduct.ironStock = ironStock;
    }
  }

  @action
  async deleteIronStockProduct(id: number) {
    try {
      const cachedProduct = this.products.find((sProduct) => sProduct.id === id)!;
      cachedProduct.minQuantity = -1;
      await updateProduct(cachedProduct);
      this.setIronStockProducts(this.ironStockProducts.filter((product) => product.id !== id));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    }
  }

  @action
  setActiveProduct(product: IApiProduct) {
    this.activeProduct = product;
  }

  @action
  setProducts(products: IApiProduct[]) {
    this.products = products;
  }

  @action
  setIronStockProducts(ironStockProducts: IIronStockProduct[]) {
    this.ironStockProducts = ironStockProducts;
  }

  @action
  setMissingIronStockProducts(ironStockProducts: IIronStockProduct[]) {
    this.missingIronStockProducts = ironStockProducts;
  }

  @action
  setProductItems(items: IProductItem[]) {
    this.productItems = items;
  }

  @action
  setAddProductFormData(data: IProduct) {
    this.addProductFormData = data;
  }

  @action
  resetAddProductFormData = () => {
    this.setProductItems([]);
    this.setAddProductFormData({
      name: "",
      iconPath: "",
      category: [],
      totalQuantity: 1,
      minQuantity: -1,
      unit: "",
      items: this.productItems,
      nutritionalValue: this.nutritionValueEmpty
    });
  };
}
