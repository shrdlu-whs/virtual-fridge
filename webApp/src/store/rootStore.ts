import { UiStateStore } from "./uiStateStore";
import { AccountStore } from "./accountStore";
import { RouteStore } from "./routeStore";
import { SettingStore } from "./settingStore";
import { NotificationStore } from "./notificationStore"
import { ProductStore } from "./productStore"
import { RecipeStore } from "./recipeStore"
import { ShoppingListStore } from "./shoppinglistStore"
import { CategoryStore } from "./categoryStore";


export class RootStore {
  uiStateStore: UiStateStore;
  accountStore: AccountStore;
  routeStore: RouteStore;
  notificationStore: NotificationStore;
  settingStore: SettingStore;
  productStore: ProductStore;
  recipeStore: RecipeStore;
  shoppingListStore: ShoppingListStore;
  categoryStore: CategoryStore;

  constructor() {
    this.uiStateStore = new UiStateStore(this);
    this.accountStore = new AccountStore(this);
    this.routeStore = new RouteStore(this);
    this.notificationStore = new NotificationStore(this);
    this.settingStore = new SettingStore(this);
    this.productStore = new ProductStore(this);
    this.recipeStore = new RecipeStore(this);
    this.shoppingListStore = new ShoppingListStore(this);
    this.categoryStore = new CategoryStore(this);
  }
}
