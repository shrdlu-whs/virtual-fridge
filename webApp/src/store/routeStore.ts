import { makeObservable, action, observable } from "mobx";
import { fromStream } from "mobx-utils";
import { RootStore } from "./rootStore";
import { Redirect } from "react-router-dom"

export class RouteStore {
  @observable redirectedTo: string | undefined = undefined;

  constructor(private rootStore: RootStore) {
    makeObservable(this)
  }

  @action
  redirect(path: string, tokenExpired: boolean = false) {
    if (this.redirectedTo !== path) {
      if (tokenExpired) {
        this.rootStore.notificationStore.createNotification(
          "token_expired",
          "error",
          true
        );
      }
      this.redirectedTo = path;
    }
  }

  @action
  resetRedirectedTo() {
    this.redirectedTo = undefined;
  }
}
