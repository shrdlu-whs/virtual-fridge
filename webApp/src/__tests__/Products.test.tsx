import React from "react";
import { INutritionValue, IProductItem, IProduct } from "../interfaces/froms/addProductForm.interface";
import { RootStore } from "../store/rootStore";
import { Provider } from "mobx-react"; 
import { render, screen } from '@testing-library/react'; 
import Product from '../components/products-content/productCardsContainer'
const products = require('../components/products-content/productCardsContainer')

test('Products exists', () => {
    expect(typeof products).toEqual('object');
});

test('test object of product', () => { 
    const nutritionValue: INutritionValue = {
        calories:450,
        carbohydrates:3,
        fat:150,
        fiber:0,
        nutritionScore:"E",
        protein:4,
        salt:3,
        saturatedFat:4,
        sugar:40
      };
  const product: IProduct = {
    name: "Schokolade",
    iconPath: "",
    category: [],
    totalQuantity: 1,
    minQuantity: -1,
    unit: "",
    items: [],
    nutritionalValue: nutritionValue}
; 
      expect(product).toMatchSnapshot();
});
test('render products', () => {
    const rootStore: RootStore = new RootStore();
    render(
        <Provider {...rootStore}>
            <Product />
         </Provider>);
    const button=screen.getByRole('button', {name: 'add'})
    const input=screen.getByLabelText('search');
    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();
});
