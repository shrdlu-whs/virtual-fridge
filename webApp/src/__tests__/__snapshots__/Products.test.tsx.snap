// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test object of product 1`] = `
Object {
  "category": Array [],
  "iconPath": "",
  "items": Array [],
  "minQuantity": -1,
  "name": "Schokolade",
  "nutritionalValue": Object {
    "calories": 450,
    "carbohydrates": 3,
    "fat": 150,
    "fiber": 0,
    "nutritionScore": "E",
    "protein": 4,
    "salt": 3,
    "saturatedFat": 4,
    "sugar": 40,
  },
  "totalQuantity": 1,
  "unit": "",
}
`;
